### PR TITLE
Fix receipt recovery & migration

### LINKED ISSUE #5823: Gnosis: `eth_getTransactionReceipt` missing `contractAddress`
**Description**
On the gnosis network, calls to `eth_getTransactionReceipt` on some contract creation transactions (e.g `0xdff4563e305339719107cb3db9e9043d92e6d0159319ef5d2383195c3b69b38a`) have a `null` `contractAddress` field.

**Steps to Reproduce**
```sh
curl $RPC --json '{"method":"eth_getTransactionReceipt","params":["0xdff4563e305339719107cb3db9e9043d92e6d0159319ef5d2383195c3b69b38a"],"id":1,
"jsonrpc":"2.0"}' | jq '.result'
```

**Actual behavior**
On a 1.19.2 Neithermind node:
```json
{
    "transactionHash": "0xdff4563e305339719107cb3db9e9043d92e6d0159319ef5d2383195c3b69b38a",
    "transactionIndex": "0x0",
    "blockHash": "0x580469139ad0ee63aa99dd91282710d2e856a4546737c1f7b7a54ef5cd31bf12",
    "blockNumber": "0x1b09d04",
    "cumulativeGasUsed": "0x29def3",
    "gasUsed": "0x29def3",
    "effectiveGasPrice": "0x138eca4807",
    "from": "0x0f41e71f39e8671d0e301c544007d4d90910e0b7",
    "to": null,
    "contractAddress": null,
    "logs": [],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "status": "0x1",
    "type": "0x2"
  }
```
**Expected behavior**
On an Erigon node:
```json
{
    "blockHash": "0x580469139ad0ee63aa99dd91282710d2e856a4546737c1f7b7a54ef5cd31bf12",
    "blockNumber": "0x1b09d04",
    "contractAddress": "0xa1dea36b62dae165072d914527e97fe4e2eed199",
    "cumulativeGasUsed": "0x29def3",
    "effectiveGasPrice": "0x138eca4807",
    "from": "0x0f41e71f39e8671d0e301c544007d4d90910e0b7",
    "gasUsed": "0x29def3",
    "logs": [],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "status": "0x1",
    "to": null,
    "transactionHash": "0xdff4563e305339719107cb3db9e9043d92e6d0159319ef5d2383195c3b69b38a",
    "transactionIndex": "0x0",
    "type": "0x2"
  }
```

### PR DESCRIPTION
**Pull request by emlautarom1**
Fixes #5823 

## Changes

- Use `AncientReceiptsBarrier` as lower bound when fixing receipts.
- Properly calculate `receipt.ContractAddress` by recovering sender.

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [] No

#### Notes on testing

Added unit test and also tested against user provided scenario on a running client.

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

## Remarks

It would be nice if we could create a unit/integration test for the associated ticket.

### CODE REVIEW

File: src/Nethermind/Nethermind.Blockchain/Receipts/ReceiptsRecovery.cs [hunk 1]
```diff
@@ -97,12 +97,17 @@ public void RecoverReceiptData(TxReceipt receipt)
 
                 Transaction transaction = _block.GetNextTransaction();
 
+                if (transaction.SenderAddress is null && _forceRecoverSender)
+                {
+                    transaction.SenderAddress = _ecdsa.RecoverAddress(transaction, !_releaseSpec.ValidateChainId);
+                }
+
```
Comment (emlautarom1) [inline src/Nethermind/Nethermind.Blockchain/Receipts/ReceiptsRecovery.cs:104]: `transaction.SenderAddress` is used by both `receipt.Sender` and `receipt.ContractAddress`, so the recovery step must be done before usage.

File: src/Nethermind/Nethermind.Init/Steps/Migrations/ReceiptFixMigration.cs [hunk 2]
```diff
@@ -47,14 +47,15 @@ public void Run()
                 _cancellationTokenSource = new CancellationTokenSource();
                 CancellationToken cancellationToken = _cancellationTokenSource.Token;
 
-                MissingReceiptsFixVisitor visitor = new MissingReceiptsFixVisitor(
-                    syncConfig.PivotNumberParsed,
+                MissingReceiptsFixVisitor visitor = new(
+                    syncConfig.AncientReceiptsBarrierCalc,
```
Comment (emlautarom1) [inline src/Nethermind/Nethermind.Init/Steps/Migrations/ReceiptFixMigration.cs:51]: `syncConfig.PivotNumberParsed` delegates to `syncConfig.PivotNumber`, which is **not a read-only field**. The pivot number is modified here:

https://github.com/NethermindEth/nethermind/blob/52d7addb56dc2d9d91dd0f8b5844c8d3fae2c521/src/Nethermind/Nethermind.Merge.Plugin/Synchronization/PivotUpdator.cs#L77

then, by the time that this code gets executed, the pivot number is not what the user originally configured.

---

Either way, we should use the `AncientReceiptsBarrier` value for this operation.

File: src/Nethermind/Nethermind.Blockchain/Receipts/ReceiptsRecovery.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Blockchain/Receipts/ReceiptsRecovery.cs:104]: Maybe add a simple test?

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Blockchain/Receipts/ReceiptsRecovery.cs (hunk 1 above)
Comment (emlautarom1) [reply src/Nethermind/Nethermind.Blockchain/Receipts/ReceiptsRecovery.cs:104]: Added test

Comment (emlautarom1) [reply src/Nethermind/Nethermind.Blockchain/Receipts/ReceiptsRecovery.cs:104]: Added `TryRecover_should_recover_contract_address` test

**Review by LukaszRozmej — APPROVED** [review]

**Review by benaadams — APPROVED** [review]

**Review by asdacap — APPROVED** [review]

**Pull request merged**